build/lib/optimize.ts

```diff
@@ -205,67 +201,123 @@ const DEFAULT_FILE_HEADER = [
+
+		const bundleInfoArray: VinylFile[] = [];
+		if (opts.bundleInfo) {
+			bundleInfoArray.push(new VinylFile({
+				path: 'bundleInfo.json',
+				base: '.',
+				contents: Buffer.from(JSON.stringify(result.bundleData, null, '\t'))
+			}));
+		}
+		es.readArray(bundleInfoArray).pipe(bundleInfoStream);
+	});
 
-		const result = es.merge(
-			loader(src, bundledFileHeader, bundleLoader, opts.externalLoaderInfo),
-			bundlesStream,
-			resourcesStream,
-			bundleInfoStream
-		);
+	const result = es.merge(
+		loader(src, bundledFileHeader, false, opts.externalLoaderInfo),
+		bundlesStream,
+		resourcesStream,
+		bundleInfoStream
+	);
 
-		return result
-			.pipe(sourcemaps.write('./', {
-				sourceRoot: undefined,
-				addComment: true,
-				includeContent: true
-			}))
-			.pipe(opts.languages && opts.languages.length ? processNlsFiles({
-				fileHeader: bundledFileHeader,
-				languages: opts.languages
-			}) : es.through())
-			.pipe(gulp.dest(out));
+	return result
+		.pipe(sourcemaps.write('./', {
+			sourceRoot: undefined,
+			addComment: true,
+			includeContent: true
+		}))
+		.pipe(opts.languages && opts.languages.length ? processNlsFiles({
+			fileHeader: bundledFileHeader,
+			languages: opts.languages
+		}) : es.through());
+}
+
+export interface IOptimizeCommonJSTaskOpts {
+	/**
+	 * The paths to consider for optimizing.
+	 */
+	entryPoints: string[];
+	/**
+	 * The folder to read files from.
+	 */
+	src: string;
+	/**
+	 * ESBuild `platform` option: https://esbuild.github.io/api/#platform
+	 */
+	platform: 'browser' | 'node' | 'neutral';
+	/**
+	 * ESBuild `external` option: https://esbuild.github.io/api/#external
+	 */
+	external: string[];
+}
+
+function optimizeCommonJSTask(opts: IOptimizeCommonJSTaskOpts): NodeJS.ReadWriteStream {
+	const esbuild = require('esbuild') as typeof import('esbuild');
+
+	const src = opts.src;
+	const entryPoints = opts.entryPoints;
+
+	return gulp.src(entryPoints, { base: `${src}`, allowEmpty: true })
+		.pipe(es.map((f: any, cb) => {
+			esbuild.build({
+				entryPoints: [f.path],
+				bundle: true,
+				platform: opts.platform,
+				write: false,
+				external: opts.external
+			}).then(res => {
+				const jsFile = res.outputFiles[0];
+				f.contents = Buffer.from(jsFile.contents);
+
+				cb(undefined, f);
+			});
+		}));
+}
+
+export function optimizeLoaderTask(src: string, out: string, bundleLoader: boolean, bundledFileHeader = '', externalLoaderInfo?: any): () => NodeJS.ReadWriteStream {
+	return () => loader(src, bundledFileHeader, bundleLoader, externalLoaderInfo).pipe(gulp.dest(out));
+}
+
+export interface IOptimizeTaskOpts {
+	out: string;
+	amd: IOptimizeAMDTaskOpts;
+	commonJS?: IOptimizeCommonJSTaskOpts;
+}
+
+export function optimizeTask(opts: IOptimizeTaskOpts): () => NodeJS.ReadWriteStream {
+	return function () {
+		return es.merge(
+			optimizeAMDTask(opts.amd),
+			opts.commonJS ? optimizeCommonJSTask(opts.commonJS) : es.through()
```

Comment:
`es.merge( something, es.through() )` is the root cause for `yarn gulp editor-distro` failing. We could do `es.merge` over a spread array and conditionally push to the array before if necessary.

Reply:
Using spread array https://github.com/microsoft/vscode/pull/160957/commits/f5e095bc4d74b6969d55a55b35487ff048f84be5